File: android/app/src/main/java/my/ssdid/wallet/domain/recovery/institutional/InstitutionalRecoveryManager.kt
package my.ssdid.wallet.domain.recovery.institutional

import kotlinx.serialization.Serializable
import my.ssdid.sdk.domain.model.Algorithm
import my.ssdid.sdk.domain.model.Identity
import my.ssdid.wallet.domain.recovery.RecoveryManager
import java.time.Instant
import java.util.Base64

/**
 * Manages institutional (organization-assisted) recovery.
 * An organization enrolls as a recovery custodian by providing their DID.
 * During recovery, the organization signs a recovery authorization that
 * proves the user's identity claim, allowing key restoration.
 */
class InstitutionalRecoveryManager(
    private val recoveryManager: RecoveryManager,
    private val storage: InstitutionalRecoveryStorage
) {
    /**
     * Enroll an organization as a recovery custodian.
     * The organization's DID is stored alongside a copy of the recovery key
     * (encrypted to the organization's public key in a real deployment).
     *
     * @param identity The identity to protect
     * @param orgDid The organization's DID
     * @param orgName Display name for the organization
     * @param encryptedRecoveryKey Recovery key encrypted to the org's public key
     */
    suspend fun enrollOrganization(
        identity: Identity,
        orgDid: String,
        orgName: String,
        encryptedRecoveryKey: ByteArray
    ): Result<OrgRecoveryConfig> = runCatching {
        require(orgDid.startsWith("did:")) { "Invalid organization DID: $orgDid" }

        // Ensure identity has a recovery key
        val hasKey = recoveryManager.hasRecoveryKey(identity.keyId)
        require(hasKey) { "Identity must have a recovery key before enrolling an organization" }

        val config = OrgRecoveryConfig(
            userDid = identity.did,
            orgDid = orgDid,
            orgName = orgName,
            encryptedRecoveryKey = Base64.getUrlEncoder().withoutPadding().encodeToString(encryptedRecoveryKey),
            enrolledAt = Instant.now().toString()
        )
        storage.saveOrgRecoveryConfig(config)
        config
    }

    /**
     * Recover using organization-provided recovery key.
     * The organization decrypts and provides the recovery key after
     * verifying the user's identity through their own KYC process.
     *
     * @param did The DID to recover
     * @param recoveryKeyBase64 The recovery private key (provided by org after verification)
     * @param name Name for the restored identity
     * @param algorithm Algorithm for new key generation
     */
    suspend fun recoverWithOrgAssistance(
        did: String,
        recoveryKeyBase64: String,
        name: String,
        algorithm: Algorithm
    ): Result<Identity> = runCatching {
        val config = storage.getOrgRecoveryConfig(did)
            ?: throw IllegalStateException("No institutional recovery configured for $did")

        // Delegate to standard recovery manager
        recoveryManager.restoreWithRecoveryKey(did, recoveryKeyBase64, name, algorithm).getOrThrow()
    }

    /**
     * Check if institutional recovery is configured for a DID.
     */
    suspend fun hasOrgRecovery(did: String): Boolean {
        return storage.getOrgRecoveryConfig(did) != null
    }

    /**
     * Get organization recovery configuration.
     */
    suspend fun getConfig(did: String): OrgRecoveryConfig? {
        return storage.getOrgRecoveryConfig(did)
    }
}

@Serializable
data class OrgRecoveryConfig(
    val userDid: String,
    val orgDid: String,
    val orgName: String,
    val encryptedRecoveryKey: String,
    val enrolledAt: String
)

interface InstitutionalRecoveryStorage {
    suspend fun saveOrgRecoveryConfig(config: OrgRecoveryConfig)
    suspend fun getOrgRecoveryConfig(userDid: String): OrgRecoveryConfig?
    suspend fun deleteOrgRecoveryConfig(userDid: String)
}

File: sdk/ios/SsdidCore/Sources/SsdidCore/Domain/Recovery/Institutional/InstitutionalRecoveryManager.swift
import Foundation

// MARK: - Models

public struct OrgRecoveryConfig: Codable, Equatable {
    public let userDid: String
    public let orgDid: String
    public let orgName: String
    public let encryptedRecoveryKey: String
    public let enrolledAt: String
}

// MARK: - InstitutionalRecoveryManager

/// Manages institutional (organization-assisted) recovery.
/// An organization enrolls as a recovery custodian by providing their DID.
/// During recovery, the organization signs a recovery authorization that
/// proves the user's identity claim, allowing key restoration.
public final class InstitutionalRecoveryManager: @unchecked Sendable {

    private let recoveryManager: RecoveryManager
    private let defaults: UserDefaults

    public init(
        recoveryManager: RecoveryManager,
        defaults: UserDefaults = .standard
    ) {
        self.recoveryManager = recoveryManager
        self.defaults = defaults
    }

    // MARK: - Public API

    /// Enroll an organization as a recovery custodian.
    /// The organization's DID is stored alongside a copy of the recovery key
    /// (encrypted to the organization's public key in a real deployment).
    ///
    /// - Parameters:
    ///   - identity: The identity to protect
    ///   - orgDid: The organization's DID
    ///   - orgName: Display name for the organization
    ///   - encryptedRecoveryKey: Recovery key encrypted to the org's public key
    /// - Returns: The stored organization recovery configuration
    public func enrollOrganization(
        identity: Identity,
        orgDid: String,
        orgName: String,
        encryptedRecoveryKey: Data
    ) async throws -> OrgRecoveryConfig {
        guard orgDid.hasPrefix("did:") else {
            throw InstitutionalRecoveryError.invalidOrgDid(orgDid)
        }

        // Ensure identity has a recovery key
        let hasKey = await recoveryManager.hasRecoveryKey(keyId: identity.keyId)
        guard hasKey else {
            throw InstitutionalRecoveryError.missingRecoveryKey
        }

        let encodedKey = encryptedRecoveryKey
            .base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")

        let config = OrgRecoveryConfig(
            userDid: identity.did,
            orgDid: orgDid,
            orgName: orgName,
            encryptedRecoveryKey: encodedKey,
            enrolledAt: ISO8601DateFormatter().string(from: Date())
        )

        saveConfig(config, userDid: identity.did)
        return config
    }

    /// Recover using organization-provided recovery key.
    /// The organization decrypts and provides the recovery key after
    /// verifying the user's identity through their own KYC process.
    ///
    /// - Parameters:
    ///   - did: The DID to recover
    ///   - recoveryKeyBase64: The recovery private key (provided by org after verification)
    ///   - name: Name for the restored identity
    ///   - algorithm: Algorithm for new key generation
    public func recoverWithOrgAssistance(
        did: String,
        recoveryKeyBase64: String,
        name: String,
        algorithm: Algorithm
    ) async throws -> Identity {
        guard getConfig(did: did) != nil else {
            throw InstitutionalRecoveryError.notConfigured(did)
        }

        // Delegate to standard recovery manager
        return try await recoveryManager.restoreWithRecoveryKey(
            did: did,
            recoveryPrivateKeyBase64: recoveryKeyBase64,
            name: name,
            algorithm: algorithm
        )
    }

    /// Check if institutional recovery is configured for a DID.
    public func hasOrgRecovery(did: String) -> Bool {
        return getConfig(did: did) != nil
    }

    /// Get organization recovery configuration.
    public func getConfig(did: String) -> OrgRecoveryConfig? {
        let key = Self.storageKey(for: did)
        guard let data = defaults.data(forKey: key) else { return nil }
        return try? JSONDecoder().decode(OrgRecoveryConfig.self, from: data)
    }

    // MARK: - Private Helpers

    private func saveConfig(_ config: OrgRecoveryConfig, userDid: String) {
        let key = Self.storageKey(for: userDid)
        if let data = try? JSONEncoder().encode(config) {
            defaults.set(data, forKey: key)
        }
    }

    private static func storageKey(for did: String) -> String {
        "ssdid_institutional_recovery_\(did)"
    }
}

// MARK: - Errors

public enum InstitutionalRecoveryError: Error, LocalizedError {
    case invalidOrgDid(String)
    case missingRecoveryKey
    case notConfigured(String)

    public var errorDescription: String? {
        switch self {
        case .invalidOrgDid(let did):
            return "Invalid organization DID: \(did)"
        case .missingRecoveryKey:
            return "Identity must have a recovery key before enrolling an organization"
        case .notConfigured(let did):
            return "No institutional recovery configured for \(did)"
        }
    }
}
